### PR TITLE
[hail] teach tables how to HTML

### LINKED ISSUE #5663: Integrate show with jupyter
`show` should integrate with the jupyter table stuff the same way that pandas does.

### PR DESCRIPTION
**Pull request by danking**
Take a look at the docs for [`IPython.display.display`](https://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html#IPython.display.display).

I preserve the user's ability to specify a custom handler. The handler is no longer given a string but an object that has a sensible `__str__` and `__repr__`. Moreover, this object has a `_repr_html_` which Jupyter uses to display an HTML table. Detecting what frontend is being run is done by `IPython.display.display`.

I use the `_Show` shim class to avoid having tables themselves print as HTML.

I also check for terminal size and use that to pick `n` and `width`.

Should `_hl_repr` live here?

Resolves #5663, #2847

### CODE REVIEW

File: hail/build.gradle [hunk 1]
```diff
@@ -279,7 +279,7 @@ task testPython(type: Exec, dependsOn: shadowJar) {
             '--color=no',
             '-r a',
             '--html=build/reports/pytest.html',
-            '--self-contained-html',
+            '--self-contained-html', '-vv', '--maxfail=1',
```
Comment (tpoterba) [inline hail/build.gradle]: disable?

File: hail/python/hail/table.py [hunk 2]
```diff
@@ -1317,8 +1338,43 @@ def format_line(values, widths, right_align):
...
+        import cgi
+
+        def hl_format(v):
+            return cgi.escape(Table._hl_repr(v))
+
+        t = self
+        t = t.flatten()
+        fields = list(t.row)
+
+        formatted_t = t.select(**{k: hl_format(v) for (k, v) in t.row.items()})
```
Comment (tpoterba) [inline hail/python/hail/table.py]: This won't work - it'll try to call hl_format on expressions. hl_format needs to be called on strings after the `take` localizes them to python

File: hail/python/hail/table.py [hunk 3]
```diff
@@ -1317,8 +1338,43 @@ def format_line(values, widths, right_align):
...
+        if types:
+            s += format_line([cgi.escape(str(t.row[f].dtype)) for f in fields])
+        s += '</thead><tbody>'
+        for row in rows:
+            s += format_line([row[f] for f in row])
+        s += '</tbody></table>'
+
+        if has_more:
+            n_rows = len(rows)
+            s += f"<p>showing top { n_rows } { 'row' if n_rows == 1 else 'rows' }</p>\n"
```
Comment (tpoterba) [inline hail/python/hail/table.py]: we have a `plural` utility function in hail.utils.misc

File: hail/python/hail/table.py [hunk 4]
```diff
@@ -1351,7 +1407,18 @@ def show(self, n=10, width=90, truncate=None, types=True, handler=print):
         handler : Callable[[str], Any]
             Handler function for data string.
         """
-        handler(self._show(n, width, truncate, types))
+        if n is None or width is None:
+            import shutil
+            (columns, lines) = shutil.get_terminal_size((80, 10))
```
Comment (tpoterba) [inline hail/python/hail/table.py]: very nice.

File: hail/python/hail/table.py (hunk 3 above)
Comment (danking) [reply hail/python/hail/table.py]: fixed

**Comment by danking** [comment]
e.g. Jupyter. Looks like Jupyter also supports terminal size, and it chooses 40 in this case, which means Jupyter told us 50 lines, which seems a bit much. In reality you can fit about 35 HTML table lines in that browser window including the Jupiter header (which eats about 5 lines).

Screen Shot 2019-03-22 at 12 04 39 PM

**Comment by danking** [comment]
IPython
Screen Shot 2019-03-22 at 12 09 18 PM

File: hail/python/hail/table.py (hunk 2 above)
Comment (tpoterba) [reply hail/python/hail/table.py]: wait, but isn't this calling `cgi.escape` inside of a TableMapRows (select) above? How does that work?

Comment (danking) [reply hail/python/hail/table.py]: lmao nice: https://github.com/python/cpython/blob/3.5/Lib/html/__init__.py#L12

replace just works.

Comment (danking) [reply hail/python/hail/table.py]:
```python
import html
print(html.escape(hl.str("tlocus<GRCh37>"))._ir)
```
```
(Apply replace (Apply replace (Apply replace (Apply replace (Apply replace (Str "tlocus<GRCh37>") (Str "&") (Str "&amp;")) (Str "<") (Str "&lt;")) (Str ">") (Str "&gt;")) (Str "\"") (Str "&quot;")) (Str "'") (Str "&#x27;"))
```

Comment (tpoterba) [reply hail/python/hail/table.py]: I don't really like this -- I think there's nothing that prevents the `cgi` or `html` implementation from changing and breaking this, and Hail string manipulation is also *extremely* slow right now.

Can we just call `hl_format` on each string coming out in line 1367?

File: hail/python/hail/table.py [hunk 5]
```diff
@@ -1202,6 +1202,44 @@ def write(self, output: str, overwrite = False, stage_locally: bool = False,
...
+            s = hl.format('%.2e', v)
+        elif isinstance(v.dtype, hl.tarray):
+            s = "[" + hl.delimit(hl.map(Table._hl_repr, v), ",") + "]"
+        elif isinstance(v.dtype, hl.tset):
+            s = "{" + hl.delimit(hl.map(Table._hl_repr, hl.array(v)), ",") + "}"
+        elif isinstance(v.dtype, hl.tdict):
+            s = "{" + hl.delimit(hl.map(lambda x: Table._hl_repr(x[0]) + ":" + Table._hl_repr(x[1]), hl.array(v)), ",") + "}"
+        elif v.dtype == hl.tstr:
+            s = hl.str('"') + hl.expr.functions._escape_string(v) + '"'
+        elif isinstance(v.dtype, (hl.tstruct, hl.tarray)):
```
Comment (tpoterba) [inline hail/python/hail/table.py]: you didn't change this code, but this should be ttuple not tarray

**Comment by tpoterba** [comment]
doctest failure from show() changing the width, it seems

File: hail/python/hail/table.py [hunk 6]
```diff
@@ -1202,6 +1202,44 @@ def write(self, output: str, overwrite = False, stage_locally: bool = False,
...
+        elif isinstance(v.dtype, hl.tarray):
+            s = "[" + hl.delimit(hl.map(Table._hl_repr, v), ",") + "]"
+        elif isinstance(v.dtype, hl.tset):
+            s = "{" + hl.delimit(hl.map(Table._hl_repr, hl.array(v)), ",") + "}"
+        elif isinstance(v.dtype, hl.tdict):
+            s = "{" + hl.delimit(hl.map(lambda x: Table._hl_repr(x[0]) + ":" + Table._hl_repr(x[1]), hl.array(v)), ",") + "}"
+        elif v.dtype == hl.tstr:
+            s = hl.str('"') + hl.expr.functions._escape_string(v) + '"'
+        elif isinstance(v.dtype, (hl.tstruct, hl.ttuple)):
+            s = "(" + hl.delimit([Table._hl_repr(v[i]) for i in range(len(v))], ",") + ")"
```
Comment (tpoterba) [inline hail/python/hail/table.py]: needs to check for empty array - that can't be type-imputed

Comment (tpoterba) [reply hail/python/hail/table.py]: (this is your test failure)

Comment (danking) [reply hail/python/hail/table.py]: le sigh. fixed.

**Pull request merged**